pkg/k8s/completion.go

```diff
@@ -0,0 +1,171 @@
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// CommandCompletion generates CLI suggestions from resources in a given cluster
+// It uses a list of arguments and a substring from the CLI to filter suggestions.
+type CommandCompletion struct {
+	k8sAPI        *KubernetesAPI
+	namespace     string
+	allNamespaces bool
+}
+
+// NewCommandCompletion creates a command completion module
+func NewCommandCompletion(
+	k8sAPI *KubernetesAPI,
+	namespace string,
+) *CommandCompletion {
+	return &CommandCompletion{
+		k8sAPI:    k8sAPI,
+		namespace: namespace,
+	}
+}
+
+// WithAllNamespaces modifies CommandCompletion to search for
+// resource suggestions across all cluster namespaces
+func (c *CommandCompletion) WithAllNamespaces() *CommandCompletion {
+	c.allNamespaces = true
+	return c
+}
+
+// Complete accepts a list of arguments and a substring to generate CLI suggestions.
+// `args` represent a list of arguments a user has already enterd in the CLI. These
+// arguments are used for determining what resource type we'd like to receive
+// suggestions for as well as a list of resources names that have already provided.
+// `toComplete` represents the string prefix of a resource name that we'd like to
+// use to search for suggestions
+//
+// If `args` is empty, send back a list of all resource types support by the CLI.
+// If `args` has at least one or more items, assume that the first item in `args`
+// is the resource type we are trying to get suggestions for e.g. Deployment, StatefulSets.
+//
+// Complete is generic enough so that it can find suggestions for any type of resource
+// in a Kubernetes cluster. It does this by first querying what GroupVersion a resource
+// belongs to and then does a dynamic `List` query to get resources under that GroupVersion
+func (c *CommandCompletion) Complete(args []string, toComplete string) ([]string, error) {
+	ctx, cancelFn := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancelFn()
+
+	suggestions := []string{}
+	if len(args) == 0 && toComplete == "" {
+		return StatAllResourceTypes, nil
+	}
+
+	if len(args) == 0 && toComplete != "" {
+		for _, t := range StatAllResourceTypes {
+			if strings.HasPrefix(t, toComplete) {
+				suggestions = append(suggestions, t)
```

Comment:
Good point -- This would work great if the number of suggestions has an upper bound limit and we could use that initialized array in both places where we return a list of resource types or a list of resource names i.e. suggestions generated by `DynamicClient`. 

If we were to initialize suggestions with the length of `StatAllResourcesTypes` in this code section, we would then need to create a new `[]string` var to return the list of suggestions generated by the `DynamicClient` like you mentioned. To keep things consistent, I think it may be better to just append both places, even though it might not be all that performant.